src/mito2/src/memtable/merge_tree/dict.rs

```diff
@@ -14,15 +14,400 @@
+        self.num_keys += 1;
+
+        // Since we store the key twice so the bytes usage doubled.
+        metrics.key_bytes += key.len() * 2;
+        self.key_bytes_in_index += key.len();
+
+        Ok(Some(pk_index))
+    }
+
+    /// Memory size of the builder.
+    pub fn memory_size(&self) -> usize {
+        self.key_bytes_in_index
+            + self.key_buffer.buffer_memory_size()
+            + self
+                .dict_blocks
+                .iter()
+                .map(|block| block.buffer_memory_size())
+                .sum::<usize>()
+    }
+
+    /// Finishes the builder.
+    pub fn finish(&mut self) -> Option<KeyDict> {
+        if self.key_buffer.is_empty() {
+            return None;
+        }
+
+        // Finishes current dict block.
+        let dict_block = self.key_buffer.finish();
+        self.dict_blocks.push(dict_block);
+        // Takes the pk to index map.
+        let mut pk_to_index = std::mem::take(&mut self.pk_to_index);
+        // Computes key position and then alter pk index.
+        let mut key_positions = vec![0; pk_to_index.len()];
+        for (i, pk_index) in pk_to_index.values_mut().enumerate() {
+            // The position of the i-th key is the old pk index.
+            key_positions[i] = *pk_index;
+            // Overwrites the pk index.
+            *pk_index = i as PkIndex;
+        }
+
+        Some(KeyDict {
+            pk_to_index,
+            dict_blocks: std::mem::take(&mut self.dict_blocks),
+            key_positions,
+        })
+    }
+
+    /// Scans the builder.
+    pub fn scan(&self) -> DictBuilderReader {
+        let sorted_pk_indices = self.pk_to_index.values().copied().collect();
+        let block = self.key_buffer.finish_cloned();
+        let mut blocks = Vec::with_capacity(self.dict_blocks.len() + 1);
+        blocks.extend_from_slice(&self.dict_blocks);
+        blocks.push(block);
+
+        DictBuilderReader::new(blocks, sorted_pk_indices)
+    }
+}
+
+/// Reader to scan the [KeyDictBuilder].
+#[derive(Default)]
+pub struct DictBuilderReader {
+    blocks: Vec<DictBlock>,
+    sorted_pk_indices: Vec<PkIndex>,
+    /// Current offset in the `sorted_pk_indices`.
+    offset: usize,
+}
+
+impl DictBuilderReader {
+    fn new(blocks: Vec<DictBlock>, sorted_pk_indices: Vec<PkIndex>) -> Self {
+        Self {
+            blocks,
+            sorted_pk_indices,
+            offset: 0,
+        }
+    }
+
+    /// Returns true if the item in the reader is valid.
+    pub fn is_valid(&self) -> bool {
+        self.offset < self.sorted_pk_indices.len()
+    }
+
+    /// Returns current key.
+    pub fn current_key(&self) -> &[u8] {
+        let pk_index = self.current_pk_index();
+        self.key_by_pk_index(pk_index)
+    }
+
+    /// Returns current [PkIndex] of the key.
+    pub fn current_pk_index(&self) -> PkIndex {
+        assert!(self.is_valid());
+        self.sorted_pk_indices[self.offset]
+    }
+
+    /// Advances the reader.
+    pub fn next(&mut self) {
```

Comment:
Question: Can the `next` method directly return `current_pk_index` and `current_key`? Then remove the two methods `current_key` and `current_pk_index`.

Reply:
We need to manually advance the reader and read the current pk_index multiple times.